Feature/seat reservation and query test + 예약 결제 구현

JupitorCentral opened the pull request:
Test Code

JupitorCentral committed:
UserManagementService - createUser 수정

JupitorCentral committed:
사용자 생성 테스트 생성 - DB에 실제로 저장되는 persistentTest 와 @Transactional 어노테이션으로 롤백하는 그냥 Test 파일 설정

JupitorCentral committed:
ReservationBookingService - book 테스트

JupitorCentral commented on src/test/java/personnel/jupitorsendsme/pulseticket/userManagement/UserManagementServicePersistenceTest.java:
얘는 사실 중복인데, 실제로 테스트 db 에 저장되는걸 보고 싶어서 넣었습니다.

JupitorCentral committed:
ReservationBookingService - book eventProxy 선언 위치 변경

JupitorCentral committed:
Seat 테이블의 status 컬럼 타입 변경 - varchar(20) 씩이나 컬럼이 클 필요가 없음.

JupitorCentral committed:
Reservation 테이블의 ReservationStatus 컬럼 타입 변경 - varchar(20) 씩이나 컬럼이 클 필요가 없음.

JupitorCentral committed:
IllegalReservationStatusDbValueException 추가

JupitorCentral committed:
Seat 단계를 진행하는 방식 변경

JupitorCentral committed:
ReservationBookingServiceIntegrationTest 테스트 변경 (setUp 단계추가), 이를 위한 Seat eventId 컬럼 및 Event 엔티티 설저 변경

ReservationBookingRequest 생성자 패턴 builder 로 통일

JupitorCentral committed:
Foreign Constraint 에 해당하는 엔티티 필드를 읽기 전용으로 변경, F key 컬럼을 따로 설정

JupitorCentral committed:
일관성이 없는 변수설정 제거

JupitorCentral committed:
Test method 가 하나인데 굳이 @BeforeEach 쓸 필요가 있나 ?

JupitorCentral committed:
노출되지 말아야 할 내부 클래스 숨김

JupitorCentral committed:
ReservationBookingService - makeReservation 구현 변경

JupitorCentral committed:
f key 코드 삭제

JupitorCentral committed:
불필요 테스트 코드 삭제

JupitorCentral committed:
테스트 h2 + embedded redis 쓰게 설정

JupitorCentral committed:
변수명 변경 ReservationBookingRequest.java -> ReservationRequest

JupitorCentral committed:
Seat 관련된 Service 생성

JupitorCentral committed:
SeatService 생성 (ReservationQueryService 에서 Seat 관련된 메소드 추출)

SeatService - createValidSeat 테스트

Foreign Key 검증을 위한 ForeignKeyValidator 생성

JupitorCentral committed:
불필요한 ForeignKeyValidator 삭제
SeatService 에 seat 저장시 유효성 검증 추가

JupitorCentral committed:
SeatService - createValidSeat 테스트 추가

JupitorCentral committed:
SeatService - createValidSeat 테스트 주석

JupitorCentral committed:
SeatService.java -> SeatManagementService.java

JupitorCentral committed:
Seat 생성시 일어날 수 있는 오류 추가

JupitorCentral committed:
포맷팅

JupitorCentral committed:
주석 수정,
SeatManagementService - createValidUser -> createUser (어차피 valid 해야 하니)

JupitorCentral committed:
Exception message 포맷 수정

JupitorCentral committed:
Reservation, Seat Enum 타입 값 수정

JupitorCentral committed:
Exception 의 파라미터 수정 (구현 숨기기), 테스트파일 위치 변경

JupitorCentral committed:
Seat ForeignKey 위반 클래스 재정의

JupitorCentral committed:
createSeat - isValidSeat 의 의미가 애매모호 한 것 같다

JupitorCentral committed:
단어 통일성 ...

JupitorCentral committed:
테스트파일 위치 변경

JupitorCentral committed:
Event - 티켓 가격 추가

JupitorCentral committed:
결제 service 생성 & 테스트

if-else-f commented on build.gradle.kts:
```diff
@@ -28,15 +30,27 @@ dependencies {
+    testImplementation("com.github.codemonstur:embedded-redis:1.4.3")
+
+    // Mockito agent (Java 21+ 지원)
+    mockitoAgent("org.mockito:mockito-core") { isTransitive = false }
```
자바 21을 아직 안써봐서 처음 보네요 ㅎㅎ
그런데 자바 21의 특징은 모두 파악하고 쓰시는거죠~?

JupitorCentral replied:
맨처음엔 프로젝트 시작했을때 세련되면서 장기적으로 유지되는 안정적인 버전 써보자 해서 21을 선택했는데,
멘토님 덕분에 java 21 의 특징을 다시 한번 짚고 넘어가게 됩니다... 하하

if-else-f commented on build.gradle.kts:
```diff
@@ -54,6 +68,20 @@ tasks.withType<JavaCompile> {
+    systemProperty("TEST_POSTGRES_USER", findProperty("TEST_POSTGRES_USER") ?: "")
+    systemProperty("TEST_POSTGRES_PASSWORD", findProperty("TEST_POSTGRES_PASSWORD") ?: "")
+    systemProperty("TEST_REDIS_PORT", findProperty("TEST_REDIS_PORT") ?: "")
+    systemProperty("TEST_REDIS_PASSWORD", findProperty("TEST_REDIS_PASSWORD") ?: "")
```
KOTLIN!!

JupitorCentral replied:
네 이거 intelliJ 에서 gradle 로 test 실행하려다가 환경변수 안먹길래 이렇게 두었는데,
지금은 intelliJ 에 내장된 JUnit 으로 테스트하고 있어서, 지금은 필요 없을 것 같네요... ㅇ.ㅇ

if-else-f commented on db_creation_query.sql:
```diff
@@ -5,20 +5,21 @@
-    password_hash VARCHAR(255)       NOT NULL,
-    created_at    TIMESTAMP          NOT NULL,
-    updated_at    TIMESTAMP          NOT NULL
+    login_id      VARCHAR(100) UNIQUE NOT NULL,
```
길이가 부족하셨나요?? ㅋㅋ

JupitorCentral replied:
test login id 설정하다보니... 😂

if-else-f commented on db_creation_query.sql:
```diff
@@ -5,20 +5,21 @@
+    id           BIGSERIAL PRIMARY KEY,
+    name         VARCHAR(255)   NOT NULL,
+    total_seats  INT            NOT NULL,
+    ticket_price Numeric(10, 2) NOT NULL,
```
indent가 변경되니 딱 수정된 부분만 찾기가 어렵네요

JupitorCentral replied:
아 이게 말씀해주신 Naver 핵데이 코딩  컨벤션 적용했던게 여기에도 적용되다보니 이렇게 됬네요.
intelliJ 에서 저장할때마다 수정하라고 되어 있어서.. 근데 java 인데 여기에도 적용됬네요 -_-;; 음 수정해야겠네요

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/controller/ReservationQueryController.java:
```diff
@@ -63,7 +65,7 @@ public Boolean isSpecificSeatAvailable(@ModelAttribute ReservationBookingRequest
 	 */
 	@GetMapping("inquiryUserReservations")
-	public List<ReservationQueryResponse> inquiryUserReservations(@ModelAttribute ReservationBookingRequest request) {
+	public List<ReservationQueryResponse> inquiryUserReservations(@ModelAttribute ReservationRequest request) {
```
`@ModelAttribute` 가 왜 필요한가요?? 그냥 지나갔네요

JupitorCentral replied:
사실 그냥 필요없었는데 이 파라미터는 url 값이 객체에 전달된다 라는걸 좀 뭔가 표현하려고 했었던 것 같아요 저때는. 하지만 less code less bug 니까, 지우는게 좋아보이네요 ㅎㅎ

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/entity/Reservation.java:
```diff
@@ -47,25 +45,40 @@ public class Reservation extends BaseEntity {
 	 */
 	@Enumerated(EnumType.STRING)
-	@Column(nullable = false, length = 20)
+	@Column(length = 20, nullable = false)
```
순서만 바뀌었네요;

JupitorCentral replied:
😅

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/entity/Seat.java:
```diff
@@ -101,20 +107,30 @@ public void sold() {
-		if (this.status == null) {
-			this.status = SeatStatus.AVAILABLE;
+	@Getter
+	public static class InvalidEventForeignKeyException extends RuntimeException {
```
Exception들은 따로 분리하는게 어떨까요??
또한, BookingException 과 같은 서비스용 exception을 만들어서 사용하면 좋을것 같아요

JupitorCentral replied:
아 저게 Seat -> Event 관계가 Seat 에만 존재하는 F key 관계다 보니, 저 클래스에 적었던 것으로 기억합니다.
그래도 역시 Exception 을 따로 두는게 좋겠죠...

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/exception/InvalidStatusException.java:
```diff
@@ -0,0 +1,25 @@
+import personnel.jupitorsendsme.pulseticket.entity.Seat;
+
+@Getter
+public class InvalidStatusException extends RuntimeException {
```
보통 `BookingException` 은 서비스용 exception을 만들고

```java
new BookingException(ResponseStatus.INVALID_STATUS, HttpStatus.BAD_REQUEST)
```

와 같이 쓰면 enum으로 에러 코드도 명시하고, 알맞은 응답 코드를 반환할 수 있어 좋습니다~

JupitorCentral replied:
오오... 알겠습니다
아 응답코드를 수정하면서 들어가는군요. 아하...
이렇게 하면 한번에 묶을수가 있겠네요.

근데 저걸 여러군데에 쓰면, logging 을 했을때 log 보고 바로 좀 더 디테일한 내용은 알 수 없을 것 같아서요... 지금 Exception 들이 좀 과한가요?

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/exception/user/UserNotFoundException.java:
```diff
@@ -1,13 +1,14 @@
-	public UserNotFoundException(String loginId) {
-		super("사용자 조회 안됨 : " + loginId);
-		this.loginId = loginId;
+	public UserNotFoundException(ReservationRequest request) {
```
오 매우 객체지향적!

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/repository/ReservationRepository.java:
```diff
@@ -1,15 +1,21 @@
+
+	@Query("SELECT r FROM Reservation r "
+		+ "LEFT JOIN FETCH r.event "
+		+ "LEFT JOIN FETCH r.seat "
```
왜 jpql을 사용하셨나요? +_+ (두근)

JupitorCentral replied:
사연이 꽤나 복잡한데요. 😂 사실은 원래는 쓸일이 없었습니다. 안쓰고 싶기도 했는데

Reservation.reserve 의 사용을 객체지향적으로 바꾸기 위해서 User, Seat 만을 파라미터로 받은다음 저장하려고 하자,
Reservation 의 Event 객체가 비어있어서 불가능하다는 에러를 마주했습니다.

```java
Reservation.reseve(User user, Seat seat) 
```
을 호출할때
Resevation 객체를 저장할때 key 만 있으면 되는거 아냐? 라고 생각해서

```java
public static Reservation reserve(User user, Seat seat) {
return Reservation.builder()
	.user(user)
	.seat(seat)
	.event(seat.getEvent())
	.status(Reservation.ReservationStatus.PENDING)
	.expiresAt(LocalDateTime.now().plus(Reservation.RESERVATION_EXPIRATION))
	.build();
}
```
에서
```java
public static Reservation reserve(User user, Seat seat) {
return Reservation.builder()
	.userId(user.getId())
	.seatId(seat.getId())
	.eventId(seat.getEventId())
	.status(Reservation.ReservationStatus.PENDING)
	.expiresAt(LocalDateTime.now().plus(Reservation.RESERVATION_EXPIRATION))
	.build();
}
```
로 바꾸었죠.

그러더니 예상치 못한 문제가 발생하였고, 그걸 해결하기 위해
jpql 을 써보았는데, 소용이 없더군요. 

해당 이슈에 대한 내용은 

wiki 에 적어두었습니다.

https://github.com/f-lab-edu/PulseTicket/wiki/JPA---%EC%98%81%EC%86%8D%EC%84%B1-%EC%BB%A8%ED%85%8D%EC%8A%A4%ED%8A%B8-%EC%BA%90%EC%8B%B1

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/EventManagementService.java:
```diff
@@ -0,0 +1,19 @@
+public class EventManagementService {
+
+	// TODO : 테스트, 예외 처리 필요 -> event, request 유효한지
+	boolean isPaymentAmountEnough(Event event, ReservationRequest request) {
```
public 으로 만들면 어떨까요??
`@Transactional` 은 public 에만 걸리는걸로 알고있어요~

JupitorCentral replied:
웁스, 깜빡했네요;

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/entity/Seat.java:
```diff
@@ -77,22 +81,24 @@ public class Seat extends BaseEntity {
-		switch (this.status) {
-			case AVAILABLE -> throw new IllegalStateException("예약 가능한 좌석 상태. 예약이 되고 나서 결제해야 함");
-			case SOLD -> throw new IllegalStateException("이미 판매됨");
+	public void proceed() {
```
`reserve`, `sold` 와 같은 명시적인 이름에서 범용적인 이름인 `proceed` 로 변경된 이유가 궁금해요~

JupitorCentral replied:
책에서 levelup 하는 부분을 본 적이 있습니다.
그래서 저한테도 적용을 하면 좋을까 생각했어요.
적용 가능하냐고 했을때 level 1 -> level 2 이런식으로 하나의 방향으로 변경할때 그게 가능한데
Reservation 의 상태변화는 그렇지 않고,Seat 은 가능하다고 생각이 들어서 이렇게 해봤습니다.

JupitorCentral replied:
아 근데 테스트코드 보니까 헷갈리네요. 다시 바꿔야지...

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
```
음.. 역시 `@Transaction` 과 `@Transaction(readOnly = true)` 가 함께 사용될 땐
각각 메소드에 annotation을 붙이는게 좋은것 같아요.

왜냐하면 어디는 있고, 어디는 없어서 일관성이 좀 깨지는것 같아요.
사실 클래스에 `@Transactional` 을 붙여서 잘 사용하진 않았던거 같아요.

JupitorCentral replied:
저도 생각해보니 메서드별로 하는게 더 명시적이고 좋은 것 같네요 ㅎㅎ

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+	 * @return Textual Diagram <br>
+	 */
+	public List<SeatStatusResponse> statusOfSeatsOfTheEvent(ReservationRequest request) {
+		List<Seat> seats = seatRepository.findByEvent_Id(request.getEventId());
```
결과가 없으면 어떻게 되나요??

JupitorCentral replied:
빈 리스트를 반환하는 것으로 알고 있습니다 :)

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+	public Seat getAvailableSeat(ReservationRequest request) {
+		Seat seat = getSeat(request);
+
+		if (seat.getStatus() != Seat.SeatStatus.AVAILABLE)
```
`seat.isUnavailable()` 과 같이 주요 로직이 model로 들어가면 더 객체지향적인 코드가 되고,
테스트를 작성하기에더 좋은것 같아요

JupitorCentral replied:
음! 아 좋네요 알겠습니다!

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+				() -> new Seat.InvalidEventForeignKeyException(seat));
+
+		Integer seatNumber = seat.getSeatNumber();
+		if (seatNumber == null || seatNumber < 1 || seatNumber > event.getTotalSeats()) {
```
`seat.isOutOfRange()` 와 같은 함수를 만들면 더 가독성이 좋을것 같은데 어떻게 생각하시나요~?

JupitorCentral replied:
정말 좋은 생각이십니다 ㅠㅠ 그 생각을 못했네요

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+	 */
+	@Transactional
+	public Seat createSeat(Seat seat) {
+		Seat created = null;
```
갑자기..??

JupitorCentral replied:
나중에 필요할 것 같기도 하고, 테스트 코드 만들어보고 싶어서 한번...만들어봤습니다 ㅎㅎ;

if-else-f commented on src/main/java/personnel/jupitorsendsme/pulseticket/service/SeatManagementService.java:
```diff
@@ -0,0 +1,107 @@
+			throw new SeatNumberDuplicateException(seat);
+		}
+
+		created = seatRepository.save(seat);
```
```java
return seatRepository.save(seat);
```

와 같이 작성하면 불필요한 변수를 제거할 수 있을것 같아요.

JupitorCentral replied:
아, 그렇네요...

if-else-f commented on src/test/java/personnel/jupitorsendsme/pulseticket/intergration/PayManagementServiceTest.java:
```diff
@@ -0,0 +1,236 @@
+@Import({PayManagementService.class, HashingServiceArgon2id.class, ReservationQueryService.class,
+	UserManagementService.class, EventManagementService.class, TestEntityManager.class})
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class PayManagementServiceTest {
```
여기에 `@Transactional` 을 붙이면 어떻게 될까요??

JupitorCentral replied:
@DataJpaTest 어노테이션 안에 이미 @Transactional 이 있기때문에, 동작상의 차이가 없다고 생각합니다.

if-else-f commented on src/test/java/personnel/jupitorsendsme/pulseticket/intergration/PayManagementServiceTest.java:
```diff
@@ -0,0 +1,236 @@
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+@Import({PayManagementService.class, HashingServiceArgon2id.class, ReservationQueryService.class,
+	UserManagementService.class, EventManagementService.class, TestEntityManager.class})
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
요건 무엇을 하는걸까요??

JupitorCentral replied:
@DataJpaTest 는 Application Context 를 생성하지 않고 Repository 쪽 의존성만 가져오는데, 이때 필요한 다른 의존성이 필요할 경우 @Import 안에 선언을 해야 의존성을 가져올 수 있다고 하더라구요.

단일 테스트 클래스 실행시에 Application Context 가 실행되서 모든 Bean 이 생성될 필요가 없다고 생각해서 이 방법을 써봤습니다.

실제로 ms 단위로 50%~30% 까지 속도차가 난걸 intelliJ 에서 확인했습니다.

JupitorCentral replied:
@TestInstance 어노테이션은 테스트 클래스의 생명주기를 변경할 수 있는 어노테이션인데, 
이 테스트클래스에서 @BeforeAll 을 사용해서 모든 테스트 메서드에서 공용으로 사용할 수 있는
테이블과 테스트데이터를 쓸려고 어노테이션을 적용했었습니다.
근데 나중에 알고보니 BeforeEach 를 쓰는게 독립적이기도 하고, 지금 제가 하는 테스트 데이터들이 그리 크지가 않아서
@BeforeAll 을 지웠는데 이 어노테이션은 깜빡하고 안지웠었네요.

if-else-f commented on src/test/java/personnel/jupitorsendsme/pulseticket/intergration/PayManagementServiceTest.java:
```diff
@@ -0,0 +1,236 @@
+	 * 예약 결제 - 유효하지 않는 사용자
+	 */
+	@Test
+	void payReservation_userNotFound() {
```
```java
@DisplayName("유효하지 않은 사용자가 예약 결제를 진행하면 UserNotFoundException를 반환한다")
```

와 같이 

1. `@DisplayName` 을 사용하여 무엇을 테스트하는지 명시하면 더 좋을것 같아요.
2. "xxx이면 yyy를 반환한다." 와 같이 더 구체적으로 작성하면 좋겠어요

if-else-f commented on src/test/java/personnel/jupitorsendsme/pulseticket/intergration/PayManagementServiceTest.java:
```diff
@@ -0,0 +1,236 @@
+	}
+
+	/**
+	 * 예약 결제 기본 테스트
```
[위의 리뷰](https://github.com/f-lab-edu/PulseTicket/pull/9/files#r2577719371) 를 같이 반영하면 좋겠어요

if-else-f commented on src/test/java/personnel/jupitorsendsme/pulseticket/intergration/ReservationBookingServiceIntegrationTest.java:
```diff
@@ -0,0 +1,76 @@
+	private final SeatRepository seatRepository;
+
+	/**
+	 * 예약 테스트
```
예약이 성공하면 어떻게 되는지 설명이 구체적으로 추가되면 좋겠어요~

if-else-f commented:
테스트 코드가 생겼네요!!

JupitorCentral committed:
불필요한 설정 코드 삭제

db_creation_query.sql indent 조정

@ModelAttribute 어노테이션 삭제

Seat->Event f key 위반 Exception 위치 변경

JupitorCentral committed:
InvalidStatusException.java -> InvalidReservationStatusException

모든 Exception 에 적당한 HttpStatus 설정

JupitorCentral committed:
f key 관계에 대해, key 를 따로 저장하는것이 아니라, 엔티티 그대로 보관하는 것으로 변경

그리고, F key 관계를 가지는 엔티티에 대해서
F key constraint 가 생성되지 않도록 설정
(foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT)

JupitorCentral committed:
@Transactional 적용을 위해
 메서드에 public modifier 추가, @Transactional annotation 을 메서드로 위치 변경

JupitorCentral committed:
객체지향화를 위한 메서드 추가

JupitorCentral committed:
불필요한 변수 정의 삭제

JupitorCentral committed:
proceed -> reserve, sell

JupitorCentral committed:
proxy 테스트 추가.

JupitorCentral committed:
테스트 버그 수정, @DisplayName 설정

if-else-f commented on .gitignore:
```diff
@@ -47,12 +47,15 @@ CLAUDE.md
 AGENTS.md
 .codex
+GEMINI.md
+.gemini
```
ㅇㄹㄴㄹㄴㅇㄹ

JupitorCentral committed:
Custom Exception 에 ProblemDetail 적용

JupitorCentral committed:
no!

JupitorCentral committed:
Integration Test 위치 변경 (Unit Test가 아닌것들),
PayManagementController 테스트 코드 추가

JupitorCentral committed:
Custom Exception 을 위한 GlobalExceptionHandler 의 잘못된 시그니쳐 수정

PayManagementControllerTest 의 Exception 발생시 올바른 ProblemDetail 형식의 Json 값을 리턴하는지에 대한 테스트코드 추가

JupitorCentral committed:
중복 테스트 코드 generic 을 사용해서 줄이기

JupitorCentral committed:
ResponseEntityExceptionHandler 를 쓸 필요가 없었다.

JupitorCentral committed:
내가 왜 썼는지 기억이 났다. Spring 이 제공하는 기본 핸들러를 사용하기 위해서이다.

JupitorCentral committed:
github action 설정 (build, test)

JupitorCentral committed:
ReservationBookingControllerTest 테스트코드 추가

JupitorCentral committed:
테스트 코드 수정
